src/wireviz/svgembed.py

```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import re
+import base64
+from pathlib import Path
+from typing import Union
+
+mime_subtype_replacements = {'jpg': 'jpeg', 'tif': 'tiff'}
+
+def embed_svg_images(svg_in: str, base_path: Union[str, Path] = Path.cwd()) -> str:
+    # first, find any image references in SVG data, and cache the respective base64-encoded image
+    images_b64 = {}  # cache of base64-encoded images
+    re_xlink=re.compile(r'xlink:href="(?P<URL>.*?)"', re.IGNORECASE)
+    for xlink in re_xlink.finditer(svg_in):
+        imgurl = xlink.group('URL')
+        if not imgurl in images_b64:  # only encode/cache every unique URL once
+            imgurl_abs = (Path(base_path) / imgurl).resolve()
+            images_b64[imgurl] = base64.b64encode(imgurl_abs.read_bytes()).decode('utf-8')
+    # second, replace links with the base64-encoded data
+    svg_out = svg_in
+    for url, b64 in images_b64.items():
+        svg_out = svg_out.replace(url, f'data:image/{get_mime_subtype(url)};base64, {b64}')
+
+    return svg_out
```

Comment:
To reduce the probability of false replacements, I suggest searching for the full `<image>` tag including any attributes preceding and/or succeeding `xlink:href`, and use `sub()` to replace all such URLs in one pass instead of separate loops for find and replace:
```suggestion
    images_b64 = {}  # cache of base64-encoded images

    def image_tag(pre: str, url: str, post: str) -> str:
        return f'<image{pre} xlink:href="{url}"{post}>'

    def replace(match: re.Match) -> str:
        imgurl = match['URL']
        if not imgurl in images_b64:  # only encode/cache every unique URL once
            imgurl_abs = (Path(base_path) / imgurl).resolve()
            images_b64[imgurl] = base64.b64encode(imgurl_abs.read_bytes()).decode('utf-8')
        return image_tag(
            match['PRE'] or '',
            f'data:image/{get_mime_subtype(imgurl)};base64, {images_b64[imgurl]}',
            match['POST'] or '')

    pattern = re.compile(
        image_tag(r'(?P<PRE> [^>]*?)?', r'(?P<URL>[^"]*?)', r'(?P<POST> [^>]*?)?'), re.IGNORECASE)
    return pattern.sub(replace, svg_in)
```

Reply:
This looks promising.
I'll need a bit of time to wrap my head around the logic; passing three `r''` strings to a function, and then passing the entire `replace` function to `pattern.sub()`...
It's a bit hard to understand; is that the most pythonic way to do it? I'm not complaining.. the overall approach of filtering the `<image>` tag and only replacing `xlink:href` within it seems like a very good solution.

Reply:
> This looks promising.
> I'll need a bit of time to wrap my head around the logic; passing three `r''` strings to a function,

If you see an alternative way to avoid repeating the `<image... xlink:href="..."...>` tag string when needing it both in the search pattern and in the replacement string, please suggest something that might be better. Maybe renaming some identifiers might help the readability? I agree rhe `pre` and `post` names are not very good.

> and then passing the entire `replace` function to `pattern.sub()`...

Using a replacement function instead of a replacement string is the thing to do when needing some processing of the matching string to create the replacement.

> It's a bit hard to understand; is that the most pythonic way to do it?

What is easy to understand might depend on what you are used to see. What is most Pythonic might also differ a bit between developers, I guess. I know there are PEPs describing what to prefer in many situations, but they cannot cover everything.

> I'm not complaining.. the overall approach of filtering the `<image>` tag and only replacing `xlink:href` within it seems like a very good solution.

Thank's